Source/WebKit/NetworkProcess/NetworkDataTaskDataURL.cpp

```diff
@@ -0,0 +1,199 @@
+{
+    cancel();
+    m_state = State::Completed;
+}
+
+NetworkDataTask::State NetworkDataTaskDataURL::state() const
+{
+    return m_state;
+}
+
+void NetworkDataTaskDataURL::setPendingDownloadLocation(const String& filename, SandboxExtension::Handle&& sandboxExtensionHandle, bool allowOverwrite)
+{
+    NetworkDataTask::setPendingDownloadLocation(filename, WTFMove(sandboxExtensionHandle), allowOverwrite);
+    m_allowOverwriteDownload = allowOverwrite;
+}
+
+String NetworkDataTaskDataURL::suggestedFilename() const
+{
+    if (!m_suggestedFilename.isEmpty())
+        return m_suggestedFilename;
+
+    String suggestedFilename = m_response.suggestedFilename();
+    if (!suggestedFilename.isEmpty())
+        return suggestedFilename;
+
+    return PAL::decodeURLEscapeSequences(m_response.url().lastPathComponent());
+}
+
+void NetworkDataTaskDataURL::didDecodeDataURL()
+{
+    ASSERT(m_state == State::Running);
+    ASSERT(m_decodingState == DecodingState::Finished);
+    if (!m_dataURLResult) {
+        if (m_client)
+            m_client->didCompleteWithError(internalError(firstRequest().url()));
+        return;
+    }
+
+    m_response = ResourceResponse::dataURLResponse(firstRequest().url(), m_dataURLResult.value());
+
+    didReceiveResponse(ResourceResponse(m_response), NegotiatedLegacyTLS::No, PrivateRelayed::No, [this, protectedThis = Ref { *this }](PolicyAction policyAction) {
+        if (m_state == State::Canceling || m_state == State::Completed)
+            return;
+
+        switch (policyAction) {
+        case PolicyAction::Use:
+            // Should not be reached for data URLs as they are normally handled in the web process.
+            invalidateAndCancel();
+            ASSERT_NOT_REACHED();
+            break;
+        case PolicyAction::Ignore:
+            invalidateAndCancel();
+            break;
+        case PolicyAction::Download:
+            downloadDecodedData();
+            break;
+        case PolicyAction::StopAllLoads:
+            ASSERT_NOT_REACHED();
+            break;
+        }
+    });
+}
+
+void NetworkDataTaskDataURL::downloadDecodedData()
+{
+    FileSystem::PlatformFileHandle downloadDestinationFile = FileSystem::openFile(m_pendingDownloadLocation, FileSystem::FileOpenMode::Truncate, FileSystem::FileAccessPermission::All, !m_allowOverwriteDownload);
+    if (!FileSystem::isHandleValid(downloadDestinationFile)) {
+#if USE(CURL)
+        ResourceError error(CURLE_WRITE_ERROR, m_response.url());
+#elif USE(SOUP)
+        ResourceError error(downloadDestinationError(m_response, "Cannot write destination file."_s));
+#endif
+        if (m_client)
+            m_client->didCompleteWithError(error);
+        invalidateAndCancel();
+        return;
+    }
+
+    auto& downloadManager = m_session->networkProcess().downloadManager();
+    auto download = makeUnique<Download>(downloadManager, m_pendingDownloadID, *this, *m_session, suggestedFilename());
+    auto* downloadPtr = download.get();
```

Comment:
What's the reason you want to use a variable `downloadPtr`. You want to avoid extra cost of `unique_ptr::operator->`?
I think typical WebKit code uses `download->`.

Reply:
I think this is because after doing WTFMove(download) you can no longer access download.

Reply:
Ah, existing code does same.
However, we have a rule to inhibit using a raw pointer of local variable. It'd nice we will have a solution for this problem.
https://github.com/WebKit/WebKit/wiki/Smart-Pointer-Usage-Guidelines